src/core/refiner.ml

```diff
@@ -0,0 +1,316 @@
+      let default () =
+        let t', a' = infer ctx te in
+        coerce ctx t' a' ty
+      in
+      match unfold te with
+      | Abst (dom, t) -> (
+          match Eval.whnf ctx ty with
+          | Prod (e1, e2) ->
+              let dom, _ = type_enforce ctx dom in
+              unif ctx dom e1;
+              let x, t, e2 = Bindlib.unbind2 t e2 in
+              let ctx = (x, dom, None) :: ctx in
+              let t = force ctx t e2 in
+              Abst (dom, Bindlib.(lift t |> bind_var x |> unbox))
+          | _ -> default () )
+      | LLet (t_ty, t, u) ->
+          let t_ty, _ = type_enforce ctx t_ty in
+          let t = force ctx t t_ty in
+          let x, u, ctx = Ctxt.unbind ctx t_ty (Some t) u in
+          let ty = Bindlib.(lift ty |> bind_var x |> unbox) in
+          let ty = Bindlib.subst ty t in
+          let u = force ctx u ty in
+          let u = Bindlib.(lift u |> bind_var x |> unbox) in
+          LLet (t_ty, t, u)
+      | _ -> default ()
+
+    and infer : ctxt -> term -> term * term =
+     fun ctx t ->
+      if !Debug.log_enabled then log "Infer [%a]" Print.pp_term t;
+      match unfold t with
+      | Patt _ -> assert false
+      | TEnv _ -> assert false
+      | Kind -> assert false
+      | Wild -> assert false
+      | TRef _ -> assert false
+      | Type -> (Type, Kind)
+      | Vari x ->
+          let a = try Ctxt.type_of x ctx with Not_found -> assert false in
+          (t, a)
+      | Symb s -> (t, !(s.sym_type))
+      (* All metavariables inserted are typed. *)
+      | (Meta (m, ts)) as t ->
+          let rec ref_esubst i dom =
+            (* Refine terms of the explicit substitution. *)
+            if i >= Array.length ts then dom else
+            match unfold dom with
+            | Prod(ai, b) ->
+                ts.(i) <- force ctx ts.(i) ai;
+                let b = Bindlib.subst b ts.(i) in
+                ref_esubst (i + 1) b
+            | _ ->
+                (* Meta type must be a product of arity greater or equal to
+                   the environment *)
+                assert false
+          in
+          let range = ref_esubst 0 !(m.meta_type) in
+          (t, range)
+      | LLet (t_ty, t, u) ->
+          (* Check that [a] is a type, and refine it. *)
+          let t_ty, _ = type_enforce ctx t_ty in
+          (* Check that [t] is of type [a'], and refine it *)
+          let t = force ctx t t_ty in
+          (* Unbind [u] and get new context with [x: t_ty ≔ t] *)
+          let x, u, ctx = Ctxt.unbind ctx t_ty (Some t) u in
+          (* Infer type of [u'] and refine it. *)
+          let u, u_ty = infer ctx u in
+          let u_ty = Bindlib.(u_ty |> lift |> bind_var x |> unbox) in
+          let u_ty = Bindlib.subst u_ty t in
+          let u = Bindlib.(u |> lift |> bind_var x |> unbox) in
+          (LLet (t_ty, t, u), u_ty)
+      | Abst (dom, b) ->
+          let dom, _ = type_enforce ctx dom in
```

Comment:
type_enforce enforces dom to be typable by a sort while, here, dom should be of type Type. Why not enforcing dom to be of type Type?

Reply:
Idem on lines 133, 141 and 203.

Reply:
I'm not sure that `let x : T := t in u` requires `T` to be of type `Type`, we may have `let Term_0 : term -> TYPE := Term z in ...`.

Reply:
Right. Some tests are necessary though, to make sure that it does not break some invariant somewhere.